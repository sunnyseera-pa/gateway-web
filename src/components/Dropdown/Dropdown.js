/** @jsx jsx */
import { cx } from '@emotion/css';
import { jsx } from '@emotion/react';
import PropTypes from 'prop-types';
import { Dropdown as BootstrapDropdown } from 'react-bootstrap';
import { ReactComponent as CheckIcon } from '../../images/check.svg';
import Icon from '../Icon';
import LayoutBox from '../LayoutBox';
import { PROP_TYPES_LAYOUTBOX } from '../LayoutBox/LayoutBox.propTypes';
import * as styles from './Dropdown.styles';

const Dropdown = ({
    defaultValue,
    value,
    options,
    variant,
    iconSelected,
    className,
    mt,
    mb,
    ml,
    mr,
    width,
    minWidth,
    maxWidth,
    size,
    ...outerProps
}) => {
    const currentValue = !value ? defaultValue : value;

    const parseOption = option => {
        return typeof option === 'string'
            ? {
                  value: option,
                  label: option,
              }
            : option;
    };

    const selectedOption = parseOption(
        options.find(option => {
            return parseOption(option).value === currentValue;
        })
    );

    return (
        <LayoutBox {...{ mt, mb, ml, mr, width, minWidth, maxWidth }}>
            <BootstrapDropdown className={cx('ui-Dropdown', className)} {...outerProps} css={styles.root({ variant, size })}>
                <BootstrapDropdown.Toggle variant={variant}>{selectedOption && selectedOption.label}</BootstrapDropdown.Toggle>
                <BootstrapDropdown.Menu>
                    {options.map(option => {
                        const { value: eventKey, label } = parseOption(option);

                        return (
                            <BootstrapDropdown.Item eventKey={eventKey} key={eventKey} className='d-flex'>
                                <div className='flex-grow'>{label}</div>
                                {eventKey === currentValue && iconSelected}
                            </BootstrapDropdown.Item>
                        );
                    })}
                </BootstrapDropdown.Menu>
            </BootstrapDropdown>
        </LayoutBox>
    );
};

Dropdown.propTypes = {
    variant: PropTypes.oneOf(['primary', 'secondary']),
    ...PROP_TYPES_LAYOUTBOX,
};

Dropdown.defaultProps = {
    variant: 'primary',
    options: [],
    iconSelected: <Icon svg={<CheckIcon />} fill='green600' color='green600' ml={1} size='xl' />,
    size: 'default',
};

export default Dropdown;
