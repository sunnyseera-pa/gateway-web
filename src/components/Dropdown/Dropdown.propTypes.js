import PropTypes from 'prop-types';

export const PROP_TYPES_DROPDOWN = {
    defaultValue: PropTypes.string,
    value: PropTypes.string,
    options: PropTypes.oneOfType([
        PropTypes.arrayOf(
            PropTypes.shape({
                label: PropTypes.node,
                value: PropTypes.oneOf([PropTypes.number, PropTypes.string]),
            })
        ),
        PropTypes.arrayOf(PropTypes.string),
    ]).isRequired,
    variant: PropTypes.oneOf(['primary', 'secondary']),
    iconSelected: PropTypes.node,
    onSelect: PropTypes.func,
    size: PropTypes.oneOf(['small', 'default', 'large']),
};
