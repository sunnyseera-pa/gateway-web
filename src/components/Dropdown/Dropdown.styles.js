import { css } from '@emotion/react';

export const root = ({ variant }) => theme => {
	const {
		colors,
		components: {
			Dropdown: { variants },
		},
	} = theme;

	return css`
		.dropdown-toggle {
			position: relative;
			text-align: left;
			width: 100%;
			padding-right: 1.5rem;

			&:after {
				position: absolute;
				right: 0.25rem;
				top: 50%;
				transform: translate(-50%, -50%);
				margin: 0;
			}
		}

		.dropdown-toggle,
		&.show .dropdown-toggle {
			border-width: 2px;
			border-style: solid;
			background: ${colors[variants[variant].background]};
			border-color: ${colors[variants[variant].borderColor]};
			color: ${colors.grey700};
		}

		.dropdown-toggle:focus {
			box-shadow: none;
			border-color: ${colors.green700};
		}

		.dropdown-menu {
			min-width: 100%;
		}
	`;
};
