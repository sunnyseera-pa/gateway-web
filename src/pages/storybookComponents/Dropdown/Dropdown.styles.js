import { css } from '@emotion/react';

export const root = ({ variant }) => theme => {
	const {
		colors,
		components: {
			Dropdown: { variants },
		},
	} = theme;

	return css`
		.dropdown-toggle,
		&.show .dropdown-toggle {
			border-width: 2px;
			border-style: solid;
			background: ${colors[variants[variant].background]};
			border-color: ${colors[variants[variant].borderColor]};
			color: ${colors.grey700};
		}

		.dropdown-toggle:focus {
			box-shadow: none;
			border-color: ${colors.green700};
		}
	`;
};
