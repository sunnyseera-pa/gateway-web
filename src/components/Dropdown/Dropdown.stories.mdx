import { Meta, Story, Canvas } from '@storybook/addon-docs';

import Dropdown from './Dropdown';

<Meta
    title='Core/Dropdown'
    component={Dropdown}
    args={{
        options: [
            {
                label: 'Recently updated',
                value: 'updated',
            },
            {
                label: 'Related resources',
                value: 'related',
            },
        ],
    }}
/>

<Canvas>
    <Story name='Variants'>
        {props => {
            return (
                <>
                    <div>
                        <div>Primary</div>
                        <Dropdown {...props} label='Primary' width='350px' mb={10} />
                    </div>
                    <div>
                        <div>Secondary</div>
                        <Dropdown {...props} label='Secondary' variant='secondary' width='350px' />
                    </div>
                </>
            );
        }}
    </Story>
    <Story name='Sizes'>
        {props => {
            return (
                <>
                    <div>
                        <div>Small</div>
                        <Dropdown {...props} size='small' label='Small' width='350px' mb={4} />
                    </div>
                    <div>
                        <div>Default</div>
                        <Dropdown {...props} size='default' width='350px' mb={4} />
                    </div>
                    <div>
                        <div>Large</div>
                        <Dropdown {...props} size='large' label='Large' width='350px' />
                    </div>
                </>
            );
        }}
    </Story>
</Canvas>
