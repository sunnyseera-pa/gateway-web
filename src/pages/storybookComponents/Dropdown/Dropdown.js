/** @jsx jsx */
import { jsx } from '@emotion/react';
import { useMemo } from 'react';
import { Dropdown as BootstrapDropdown } from 'react-bootstrap';
import * as styles from './Dropdown.styles';
import Icon from '../../storybookComponents/Icon';
import { addCommonPropTypes } from '../../../configs/propTypes';
import { cx } from '@emotion/css';
import { ReactComponent as VersionAccepted } from '../../../images/check.svg';

const Dropdown = ({ defaultValue, value, options, variant, iconSelected, className, ...outerProps }) => {
	const currentValue = !value ? defaultValue : value;

	const parseOption = option => {
		return typeof option === 'string'
			? {
					value: option,
					label: option,
			  }
			: option;
	};

	const selectedOption = useMemo(
		() =>
			parseOption(
				options.find(option => {
					return parseOption(option).value === currentValue;
				})
			),
		[currentValue]
	);

	return (
		<BootstrapDropdown className={cx('ui-Dropdown', className)} {...outerProps} css={styles.root({ variant })}>
			<BootstrapDropdown.Toggle variant={variant}>{selectedOption && selectedOption.value}</BootstrapDropdown.Toggle>
			<BootstrapDropdown.Menu>
				{options.map(option => {
					const { value: eventKey, label } = parseOption(option);

					console.log('XYZ', eventKey === currentValue);

					return (
						<BootstrapDropdown.Item eventKey={eventKey} className='d-flex'>
							<div className='flex-grow'>{label}</div>
							{eventKey === currentValue && iconSelected}
						</BootstrapDropdown.Item>
					);
				})}
			</BootstrapDropdown.Menu>
		</BootstrapDropdown>
	);
};

Dropdown.propTypes = addCommonPropTypes({});

Dropdown.defaultProps = {
	variant: 'primary',
	options: [],
	iconSelected: <Icon name='check' fill='green600' color='green600' ml={1} size='xl' />,
};

export default Dropdown;
