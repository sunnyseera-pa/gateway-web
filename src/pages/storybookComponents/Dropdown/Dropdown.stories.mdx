import { Meta, Story, Canvas } from '@storybook/addon-docs';

import Dropdown from './Dropdown';

<Meta
	title='Dropdown'
	component={Dropdown}
	args={{
		options: [
			{
				label: 'Recently updated',
				value: 'updated',
			},
			'Related resources',
		],
	}}
/>

<Canvas>
	<Story name='Basic'>
		{props => {
			const [value, setValue] = React.useState('updated');
			const handleChange = React.useCallback(value => {
				setValue(value);
			}, []);
			return <Dropdown {...props} onSelect={handleChange} value={value} width='200px' />;
		}}
	</Story>
</Canvas>
