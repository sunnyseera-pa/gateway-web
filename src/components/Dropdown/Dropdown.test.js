import { render, waitFor } from '@testing-library/react';
import React from 'react';
import Dropdown from '.';

const mockOnSelect = jest.fn();

const props = {
	onSelect: mockOnSelect,
	options: ['relevance', 'recentlyadded'],
	value: 'relevance',
	iconSelected: <span>Selected icon</span>,
};

let wrapper;

describe('Given the Dropdown component', () => {
	describe('When it is rendered', () => {
		beforeAll(async () => {
			wrapper = render(<Dropdown {...props} />, {
				wrapper: Providers,
			});

			const input = wrapper.container.querySelector('button');

			fireEvent.click(input);
		});

		it('Then matches the previous snapshot', () => {
			expect(wrapper.container).toMatchSnapshot();
		});

		it('Then has a selected icon', async () => {
			await waitFor(() => expect(wrapper.getByText('Selected icon')).toBeTruthy());
		});

		describe('And it is selected', () => {
			beforeAll(() => {
				const link = wrapper.container.querySelectorAll('a')[1];

				fireEvent.click(link);
			});

			it('Then calls onSelect', async () => {
				await waitFor(() => expect(mockOnSelect.mock.calls[0][0]).toEqual('recentlyadded'));
			});
		});
	});
});
