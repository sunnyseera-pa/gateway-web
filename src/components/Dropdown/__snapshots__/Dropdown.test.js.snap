// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the Dropdown component When it is rendered Then matches the previous snapshot 1`] = `
<div>
  <div
    class="ui-Dropdown css-1sivg29 css-57mu23-Dropdown show dropdown"
  >
    <button
      aria-expanded="true"
      aria-haspopup="true"
      class="dropdown-toggle btn btn-primary"
      type="button"
    >
      relevance
    </button>
    <div
      aria-labelledby=""
      class="dropdown-menu show"
      data-popper-escaped="true"
      data-popper-placement="bottom-start"
      data-popper-reference-hidden="true"
      style="position: absolute; top: 0px; left: 0px; margin: 0px; transform: translate(0px, 0px);"
      x-placement="bottom-start"
    >
      <a
        class="d-flex dropdown-item"
        href="#"
        role="button"
      >
        <div
          class="flex-grow"
        >
          relevance
        </div>
        <span>
          Selected icon
        </span>
      </a>
      <a
        class="d-flex dropdown-item"
        href="#"
        role="button"
      >
        <div
          class="flex-grow"
        >
          recentlyadded
        </div>
      </a>
    </div>
  </div>
</div>
`;
